File: src/update/update.cpp
#include <pybind11/pybind11.h>
#include "../wrappers/classifier/Classifier.h"
#include "../randomFerns/RandomFerns.h"

namespace py = pybind11;


// update classifier using positive smaples
void fun_update_positive_samples(const cv::Mat &img, CvRect rec, Classifier* clfr, RandomFerns* RFs, int updSamps, float varphi){
    // inputs:
    // img: image (I)
    // rec: image box
    // clfr: classifier
    // RFs: fern parameters
    // updSamps: num. new (updating) samples (Nu)
    // varphi: image shift rate

    // parameters
    int minSize = 20;  // min. image size
    float posLabel = 1.0; // sample label (y=+1)

    // variables
    CvRect newRec;  // new bounding box
    int ua,va,ub,vb;  // image coordinates
    int objHeight,objWidth;  // object heigth and width (Bu,Bv)

    // object size
    clfr->get_object_size(objHeight, objWidth);

    // updating positive samples
    for (int iter=0; iter<updSamps; iter++){

        // new rectangle coordinates
        ua = rec.y + round(varphi*rec.height*((float)rand()/RAND_MAX - 0.5));
        va = rec.x + round(varphi*rec.width*((float)rand()/RAND_MAX  - 0.5));
        ub = rec.y + rec.height + round(varphi*rec.height*((float)rand()/RAND_MAX - 0.5));
        vb = rec.x + rec.width + round(varphi*rec.width*((float)rand()/RAND_MAX  - 0.5));

        // check limits
        if (va<0){va = 0;}
        if (ua<0){ua = 0;}
        if (vb>=img.cols){vb = img.cols-1;}
        if (ub>=img.rows){ub = img.rows-1;}

        // check min. size
        if ((vb-va)>minSize && (ub-ua)>minSize){
            std::cout << "rec.y - " << rec.y << ", rec.x - " << rec.x << ", rec.height - " << rec.height << ", rec.width" << rec.width << std::endl;
            std::cout << "min size " << minSize << " vb-va " << (vb - va) << " ub - ua  "
                      << (ub - ua) << " va " << va << " ua " << ua << " vb " << vb << " ub " << ub << std::endl;

            // new bounding box
            newRec = cvRect(va, ua, vb-va, ub-ua);

            // image patch using ROI
            cv::Mat patch = cv::Mat(cvSize(objWidth, objHeight), CV_8UC3);
            cv::resize(img(newRec), patch, cvSize(objWidth, objHeight));

            // fern maps (convolve ferns with the image)
//            cv::Mat fernMaps = fern_maps(patch, RFs);

            // update classifier
//            clfr->update(fernMaps, posLabel);

        }
    }
};


void update_pos_samples(py::array_t<unsigned short int> &fern_map, float label) {
    std::cout << "update_pos_samples" << std::endl;
}

void update_neg_samples() {
    std::cout << "update_neg_samples" << std::endl;
}

void update_samples(py::module &m) {
    m.def("update_positive_samples", &update_pos_samples,
            py::arg("image"),
            py::arg("rectangle"));
    m.def("update_negative_samples", &update_neg_samples);
}

File: detector/functions/update.py
import time

import cv2
import numpy as np

import detect


def update_positive_samples(frame, rectangle, classifier, random_ferns, train_samples, varphi):
    start_time = time.time()
    min_size = 20
    pos_label = 1.0

    cols, rows = classifier.get_object_size()

    height, width, channel = frame.shape

    for i in range(train_samples):
        ua = rectangle.y + round(varphi * rectangle.height * (np.random.uniform(-0.5, 0.5)))
        va = rectangle.x + round(varphi * rectangle.width * (np.random.uniform(-0.5, 0.5)))
        ub = rectangle.y + rectangle.height + round(varphi * rectangle.height * (np.random.uniform(-0.5, 0.5)))
        vb = rectangle.x + rectangle.width + round(varphi * rectangle.width * (np.random.uniform(-0.5, 0.5)))

        if va < 0:
            va = 0
        if ua < 0:
            ua = 0
        if vb >= height:
            vb = height - 1
        if ub >= width:
            ub = width - 1

        # print('pos: va - {}, ua - {}, vb - {}, ub - {}, vb - va > min_size - {}, ub - ua > min_size - {}'
        #       .format(va, ua, vb, ub, vb - va > min_size, ub - ua > min_size))

        if vb - va > min_size and ub - ua > min_size:
            patch = cv2.resize(frame[ua: ub, va: vb], (rows, cols))
            f_maps = detect.fern_maps(patch, random_ferns)
            classifier.update(f_maps, pos_label)
        end_time = time.time() - start_time
        print('update_pos_samples time - {}'.format(end_time))
